spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,181 @@
+The method for which the invoker is built is called the _target method_ of the invoker, and the managed bean is called the _target bean_ of the invoker.
+
+Invalid target methods are:
+
+* private methods,
+* constructors,
+* methods declared on the `java.lang.Object` class, except of the `toString()` method,
+* methods that are not declared on the bean class of the target bean or inherited from its supertypes.
+
+Attempting to build an invoker for an invalid target method leads to a deployment problem.
+
+When the target bean is not a managed bean, attempting to build an invoker leads to a deployment problem.
+When the target bean is an interceptor, attempting to build an invoker leads to a deployment problem.
+
+Multiple managed beans may inherit a method from a common supertype.
+In that case, an invoker must be built for each target bean individually.
+An invoker built for one target bean may not be used to invoke the target method on an instance of another target bean.
+
+The only way to build an invoker is using the `InvokerBuilder`.
+An `InvokerBuilder` can only be obtained in CDI portable extensions and build compatible extensions.
+See <<invoker_builder>> for more information.
+
+[[using_invoker]]
+=== Using an `Invoker`
+
+The `Invoker` interface contains a single method:
+
+[source,java]
+----
+public interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments);
+}
+----
+
+This section describes the behavior of the `invoke()` method.
+
+Calling `invoke()` invokes the target method on given `instance` of the target bean, passing given `arguments`.
+
+If the target method is `static`, the `instance` is ignored; by convention, it should be `null`.
+If the target method returns normally, its return value is returned, unless the target method is declared `void`, in which case `null` is returned.
+If the target method throws an exception, it is rethrown.
+The `invoke()` method does not declare any checked exception; implementations are supposed to use the "sneaky throw" idiom.
+// TODO maybe change that?
+
+If the target method is not `static` and `instance` is `null`, a `NullPointerException` is thrown.
+If the target method is not `static` and the `instance` is not <<invoker_assignability,assignable>> to the target bean, a `ClassCastException` is thrown.
+
+Correspondence between given `arguments` and declared parameters of the target method is positional: the Nth element of the `arguments` array is passed as the Nth argument to the target method.
+If the target method is a variable arity method, the last element of the `arguments` array corresponds to the variable arity parameter (and therefore must be an array).
+
+If the target method declares no parameter, `arguments` are ignored.
+If the target method declares any parameter and `arguments` is `null`, `NullPointerException` is thrown.
+If the `arguments` array has fewer elements than the number of parameters of the target method, `ArrayIndexOutOfBoundsException` is thrown.
```

Comment:
This is one of the points where a method handles implementation won't behave equally (unless you check for this manually) - IIRC, it'd instead throw `IllegalArgumentException`.

Reply:
Yes. We'll need to figure out the best way here, based on what both our implementations can easily do, plus taking into account a possible implementation based on classic reflection.

Reply:
To be clear, I'm not exactly _sure_ about the `InvocationTargetException`. It is something I very much hate about reflection. Maybe we should do this instead:

- specify that for all these validations, a generic `RuntimeException` is thrown
- add `throws Exception` to the `Invoker.invoke()` method
- specify that exceptions thrown by the target method are rethrown

(This would leave custom throwables undefined, which would be consistent with the rest of CDI and Interceptors and everything.)

Reply:
I am not sure I understand the ups and downs of wrapping vs re-throwing in this scenario. Would you mind elaborating on that, please?

Reply:
The upside of wrapping is that callers can easily detect whether the exception was thrown by the invoked method or the invoker mechanism itself failed (presumably because the caller passed an incorrect argument to `invoke()`). The downside is that unwrapping becomes necessary, which is often a source of subtle problems.

Reply:
Got rid of `InvocationTargetException`, `Invoker.invoke()` is now supposed to rethrow target method exceptions directly.